### DISCUSSION TITLE
Audio Worklet Support (10.2, Webpack 5)

### DISCUSSION THREAD

**Discussion by thecynicalpaul** [vercel/next.js#62376] [Help]
### What version of Next.js are you using?

10.2.0

### What version of Node.js are you using?

14.15.3

### What browser are you using?

Latest Chrome and Safari (14.5)

### What operating system are you using?

Windows 10, and macOS Big Sur

### How are you deploying your application?

next start

### Describe the Bug

Our application is built on top of Next, and we are currently in the process of migrating our audio recording engine to Audio Worklet API from Script Processor. (with backwards compatibility of course) Part of the transition is also upgrading to Webpack 5. We are utilizing both Web Workers and Audio Worklets. Prior to switch to Webpack 5, which comes with native support for Workers, (or so it appears) we used [worker-plugin](https://github.com/GoogleChromeLabs/worker-plugin). It worked wonderfully for both, however with the transition we can no longer rely on it, because it uses outdated webpack's APIs, and Next comes with its own bundled Webpack 4 and 5, which don't seem to have backwards compatibility included.

Now the challenge is to get these working with Webpack 5 bundled with Next. The first issue comes from web worker's global scope being undefined. This can be resolved with by setting `config.output.globalObject` to `"(typeof self !== 'undefined' ? self : this)"`. Once workers are functional, the next issue comes when trying to bundle worklet's code. Currently, Webpack 5 doesn't support worklet import, but they expose parser config, which can be used to tell webpack to load it as a worker, as per this [Github issue](https://github.com/webpack/webpack/issues/11543#issuecomment-826938252). This will fail with Next unless you also set `config.output.publicPath = "/_next/";` Upon loading the chunk via `audioContext.audioWorklet.addModule(...)`, the code crashes with `Uncaught TypeError: Cannot read property 'webpackChunk_N_E' of undefined`. If we inspect the chunk containing bundled worklet code, we'll see that this prop is being used in the following way:
```js
var chunkLoadingGlobal = (typeof self !== 'undefined' ? self : this)["webpackChunk_N_E"] = (typeof self !== 'undefined' ? self : this)["webpackChunk_N_E"] || [];
var parentChunkLoadingFunction = chunkLoadingGlobal.push.bind(chunkLoadingGlobal);
.
.
.
(typeof self !== 'undefined' ? self : this)["webpackHotUpdate_N_E"] = function...
```
It's clear, that `AudioWorkletGlobalScope` doesn't have either `self` or `this`, so that's why it's not going too well. 

So the question is, how can this be worked around? Audio Worklets seem to be getting very little attention still from both Next and Webpack, despite it being available by default now even in _Safari_. (since 14.5)

Sources:
- https://github.com/GoogleChromeLabs/worker-plugin
- https://webpack.js.org/configuration/module/#moduleparserjavascript
- https://github.com/webpack/webpack/issues/11543
- https://github.com/vercel/next.js/tree/canary/examples/with-web-worker

### Expected Behavior

For the Audio Worklet Processor chunk to be loaded without issues and be functional when invoked by the Audio Context.

### To Reproduce

1. Setup the latest version of Next JS as normal.
2. Follow the steps to get Audio Worklets bundled.

Below are the code snippets representing our current state. (we are using Typescript)

`next.config.js`
```js
module.exports = withPlugins([...], {
  future: {webpack5: true},
  webpack: (config, options) => {
    config.output.globalObject = `(typeof self !== 'undefined' ? self : this)`;
    config.output.publicPath = "/_next/";

    config.resolve = {
      ...config.resolve,
      alias: {
        ...config.resolve.alias,
        "audio-worklet": path.resolve(__dirname, "src/util/audio-worklet")
      }
    }

    config.module.parser = {
      ...config.module.parser,
      javascript: {
        worker: ["AudioWorklet from audio-worklet", "..."]
      }
    }

    config.output.chunkFilename = options.isServer
      ? `${options.dev ? "[name].[hash]" : "[name].[chunkhash]"}.js`
      : `static/chunks/${options.dev ? "[name].[hash]" : "[name].[chunkhash]"}.js`;

    return config;
  }
});
```

`src/util/audio-worklet.ts`
```typescript
export const AudioWorklet = (url: URL) => {
  return url as unknown as string;
};
```

`.../audio-context.ts`
```typescript
import { AudioWorklet } from "audio-worklet";

const audioContext = new AudioContext();
audioContext.audioWorklet.addModule(new AudioWorklet(new URL("path/to/processor.worklet.ts", import.meta.url)));
```

`.../audio-worklet-processor.worklet.ts`
```typescript
// import statements for utility code

class Processor extends AudioWorkletProcessor {
  // Your run of the mill processor code
}

registerProcessor("processor", Processor);
```

**Comment by thecynicalpaul** [comment]
Update: after much digging, seems like this issue is caused by webpack importing scripts in the chunk. Specifically, when we try to import any non-trivial code, such as lodash, or use data structures that babel deems necessary to use its shims for, Webpack injects `installChunk` function which uses `importScripts`, which obviously is not supported by AudioWorklets. This also inserts the code above that causes the `Uncaught TypeError: Cannot read property 'webpackChunk_N_E' of undefined`. 

Here's a repo reproducing it one for one: https://github.com/thecynicalpaul/test-audio-worklet-webpack5 (`nmp i`, then `npm run dev`)

Any advice or insights would be amazing!

**Comment by jankaifer** [comment]
I quickly read through the issue and some of the related issues that are linked here.

Is this still relevant in `next@canary`?
And this looks like webpack bug so we will convert this to a discussion unless we can easily mitigate this in `next.js`.
